plasmapy/utils/decorators/tests/test_checks.py

```diff
@@ -10,10 +11,939 @@
+        """
+        # setup test cases
+        # 'setup' = arguments for `CheckUnits` and wrapped function
+        # 'output' = expected return from wrapped function
+        # 'raises' = if an Exception is expected to be raised
+        # 'warns' = if a warning is expected to be issued
+        #
+        _cases = [
+            # clean execution
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.cm,
+                                  'y': u.cm,
+                                  'checks_on_return': u.cm}},
+             'output': 5 * u.cm,
+             },
+
+            # argument fails checks
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.g,
+                                  'y': u.cm,
+                                  'checks_on_return': u.cm}},
+             'raises': u.UnitTypeError,
+             },
+
+            # return fails checks
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.cm,
+                                  'y': u.cm,
+                                  'checks_on_return': u.km}},
+             'raises': u.UnitTypeError,
+             },
+        ]
+
+        # test
+        for case in _cases:
+            wfoo = CheckUnits(**case['setup']['checks'])(case['setup']['function'])
+
+            args = case['setup']['args']
+            kwargs = case['setup']['kwargs']
+
+            if 'raises' in case:
+                with pytest.raises(case['raises']):
+                    wfoo(*args, **kwargs)
+            else:
+                assert wfoo(*args, **kwargs) == case['output']
+
+    def test_cu_preserves_signature(self):
+        """Test `CheckValues` preserves signature of wrapped function."""
+        # I'd like to directly dest the @preserve_signature is used (??)
+
+        wfoo = CheckUnits()(self.foo_no_anno)
+        assert hasattr(wfoo, '__signature__')
+        assert wfoo.__signature__ == inspect.signature(self.foo_no_anno)
+
+    @mock.patch(CheckUnits.__module__ + '.' + CheckUnits.__qualname__,
+                side_effect=CheckUnits, autospec=True)
+    def test_decorator_func_def(self, mock_cu_class):
+        """
+        Test that :func:`~plasmapy.utils.decorators.checks.check_units` is
+        properly defined.
+        """
+        # create mock function (mock_foo) from function to mock (self.foo_no_anno)
+        mock_foo = mock.Mock(side_effect=self.foo_no_anno, name='mock_foo',
+                             autospec=True)
+        mock_foo.__name__ = 'mock_foo'
+        mock_foo.__signature__ = inspect.signature(self.foo_no_anno)
```

Comment:
Does this mock serve any purpose besides  checking if it's been called?

Reply:
I could be open to an argument to remove this mock.

This mock is setup to ensure the decorator `check_units` actually calls the wrapped function. (see the `assert mock_foo.called` [below](https://github.com/rocco8773/PlasmaPy/blob/576e49bb3ef2d909833449cc9be918a28dbbb276/plasmapy/utils/decorators/tests/test_checks.py#L542))